core/src/Cabana_Experimental_NeighborList.hpp

```diff
@@ -114,36 +114,77 @@ namespace Experimental
 template <typename Tag>
-struct NeighborDiscriminatorCallback;
+struct CollisionFilter;
+
+template <>
+struct CollisionFilter<FullNeighborTag>
+{
+    KOKKOS_FUNCTION bool static keep( int i, int j ) noexcept
+    {
+        return i != j; // discard self-collision
+    }
+};
 
 template <>
-struct NeighborDiscriminatorCallback<Cabana::FullNeighborTag>
+struct CollisionFilter<HalfNeighborTag>
+{
+    KOKKOS_FUNCTION static bool keep( int i, int j ) noexcept { return i > j; }
+};
+
+// Custom callback for ArborX::BVH::query()
+template <typename Tag>
+struct NeighborDiscriminatorCallback
 {
     using tag = ArborX::Details::InlineCallbackTag;
     template <typename Predicate, typename OutputFunctor>
-    KOKKOS_FUNCTION void operator()( Predicate const &predicate, int i,
+    KOKKOS_FUNCTION void operator()( Predicate const &predicate,
+                                     int primitive_index,
                                      OutputFunctor const &out ) const
     {
-        if ( getData( predicate ) != i ) // discard self-collision
+        int const predicate_index = getData( predicate );
+        if ( CollisionFilter<Tag>::keep( predicate_index, primitive_index ) )
         {
-            out( i );
+            out( primitive_index );
         }
     }
 };
 
-template <>
-struct NeighborDiscriminatorCallback<Cabana::HalfNeighborTag>
+template <typename Counts, typename Tag>
+struct NeighborDiscriminatorCallback2D_FirstPass
 {
+    Counts counts;
     using tag = ArborX::Details::InlineCallbackTag;
     template <typename Predicate, typename OutputFunctor>
-    KOKKOS_FUNCTION void operator()( Predicate const &predicate, int i,
-                                     OutputFunctor const &out ) const
+    KOKKOS_FUNCTION void operator()( Predicate const &predicate,
+                                     int primitive_index,
+                                     OutputFunctor const & ) const
     {
-        if ( getData( predicate ) > i )
+        int const predicate_index = getData( predicate );
+        if ( CollisionFilter<Tag>::keep( predicate_index, primitive_index ) )
         {
-            out( i );
+            ++counts( predicate_index );
+        }
+    }
+};
+
+template <typename Counts, typename Neighbors, typename Tag>
+struct NeighborDiscriminatorCallback2D_SecondPass
+{
+    Counts counts;
+    Neighbors neighbors;
+    using tag = ArborX::Details::InlineCallbackTag;
+    template <typename Predicate, typename OutputFunctor>
+    KOKKOS_FUNCTION void operator()( Predicate const &predicate,
+                                     int primitive_index,
+                                     OutputFunctor const & ) const
+    {
+        int const predicate_index = getData( predicate );
+        if ( CollisionFilter<Tag>::keep( predicate_index, primitive_index ) )
+        {
+            assert( count( predicate_index ) < (int)neighbors.extent( 1 ) );
+            neighbors( predicate_index, counts( predicate_index )++ ) =
```

Comment:
If I remember correctly the fill will happen in serial per predicate. In the future I guess `counts` would need to be an atomic view if hierarchical parallelism was used.

Reply:
Yes.  I should had a word of warning here.  You are right I am taking advantage of my knowledge of how the traverse is currently implemented here.

Reply:
I added a comment.